libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,405 @@
+	// These are the source units we will sent diagnostics to the client for sure,
+	// even if it is just to clear previous diagnostics.
+	map<string, Json::Value> diagnosticsBySourceUnit;
+	for (string const& sourceUnitName: m_fileRepository.sourceUnits() | ranges::views::keys)
+		diagnosticsBySourceUnit[sourceUnitName] = Json::arrayValue;
+	for (string const& sourceUnitName: m_nonemptyDiagnostics)
+		diagnosticsBySourceUnit[sourceUnitName] = Json::arrayValue;
+
+	for (shared_ptr<Error const> const& error: m_compilerStack.errors())
+	{
+		SourceLocation const* location = error->sourceLocation();
+		if (!location || !location->sourceName)
+			// LSP only has diagnostics applied to individual files.
+			continue;
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = toDiagnosticSeverity(error->type());
+		jsonDiag["code"] = Json::UInt64{error->errorId().error};
+		string message = error->typeName() + ":";
+		if (string const* comment = error->comment())
+			message += " " + *comment;
+		jsonDiag["message"] = move(message);
+		jsonDiag["range"] = toRange(*location);
+
+		if (auto const* secondary = error->secondarySourceLocation())
+			for (auto&& [secondaryMessage, secondaryLocation]: secondary->infos)
+			{
+				Json::Value jsonRelated;
+				jsonRelated["message"] = secondaryMessage;
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	m_nonemptyDiagnostics.clear();
+	for (auto&& [sourceUnitName, diagnostics]: diagnosticsBySourceUnit)
+	{
+		Json::Value params;
+		params["uri"] = m_fileRepository.sourceUnitNameToClientPath(sourceUnitName);
+		if (!diagnostics.empty())
+			m_nonemptyDiagnostics.insert(sourceUnitName);
+		params["diagnostics"] = move(diagnostics);
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (m_state != State::ExitRequested && m_state != State::ExitWithoutShutdown && !m_client.closed())
+	{
+		MessageID id;
+		try
+		{
+			optional<Json::Value> const jsonMessage = m_client.receive();
+			if (!jsonMessage)
+				continue;
+
+			if ((*jsonMessage)["method"].isString())
+			{
+				string const methodName = (*jsonMessage)["method"].asString();
+				id = (*jsonMessage)["id"];
+
+				if (auto handler = valueOrDefault(m_handlers, methodName))
+					handler(id, (*jsonMessage)["params"]);
+				else
+					m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+			}
+			else
+				m_client.error({}, ErrorCode::ParseError, "\"method\" has to be a string.");
+		}
+		catch (exception const& _exception)
+		{
+			m_client.error(id, ErrorCode::InternalError, "Unhandled exception: "s + boost::diagnostic_information(_exception));
+		}
+		catch (...)
+		{
+			m_client.error(id, ErrorCode::InternalError, "Unhandled exception: "s + boost::current_exception_diagnostic_information());
+		}
```

Comment:
With a `catch (...)` block this is not superfluous. `diagnostic_information` does not provide any extra useful info over `current_exception_diagnostic_information()`.
```suggestion
		catch (...)
		{
			m_client.error(id, ErrorCode::InternalError, "Unhandled exception: "s + boost::current_exception_diagnostic_information());
		}
```